core/src/parameter.jl

```diff
@@ -1028,57 +927,80 @@ It is used to create views of `u`, and an low-latency alternative to making `u`
-    const level_demand::LevelDemand
-    const flow_demand::FlowDemand
-    const subgrid::Subgrid
-    # Per state the in- and outflow links associated with that state (if they exist)
-    const state_inflow_link::Vector{LinkMetadata} = LinkMetadata[]
-    const state_outflow_link::Vector{LinkMetadata} = LinkMetadata[]
+@kwdef mutable struct ParametersMutable
     all_nodes_active::Bool = false
     tprev::Float64 = 0.0
+end
+
+@kwdef struct ParametersNonDiff{C1, C2, C3, C4, C5}
+    starttime::DateTime
+    graph::ModelGraph
+    allocation::Allocation
+    basin::Basin{C1, C2}
+    linear_resistance::LinearResistance
+    manning_resistance::ManningResistance
+    tabulated_rating_curve::TabulatedRatingCurve
+    level_boundary::LevelBoundary{C3}
+    flow_boundary::FlowBoundary{C4}
+    pump::Pump
+    outlet::Outlet
+    terminal::Terminal
+    discrete_control::DiscreteControl
+    continuous_control::ContinuousControl
+    pid_control::PidControl
+    user_demand::UserDemand{C5}
+    level_demand::LevelDemand
+    flow_demand::FlowDemand
+    subgrid::Subgrid
+    # Per state the in- and outflow links associated with that state (if they exist)
+    state_inflow_link::Vector{LinkMetadata} = LinkMetadata[]
+    state_outflow_link::Vector{LinkMetadata} = LinkMetadata[]
     # Sparse matrix for combining flows into storages
-    const flow_to_storage::SparseMatrixCSC{Float64, Int64} = spzeros(1, 1)
+    flow_to_storage::SparseMatrixCSC{Float64, Int64} = spzeros(1, 1)
     # Water balance tolerances
-    const water_balance_abstol::Float64
-    const water_balance_reltol::Float64
+    water_balance_abstol::Float64
+    water_balance_reltol::Float64
     # State at previous saveat
-    const u_prev_saveat::Vector{Float64} = Float64[]
+    u_prev_saveat::Vector{Float64} = Float64[]
     # Node ID associated with each state
-    const node_id::Vector{NodeID} = NodeID[]
-    # Range per states component
-    const state_ranges::StateRanges = StateRanges()
+    node_id::Vector{NodeID} = NodeID[]
+    # Range per states or cache component
+    state_ranges::StateRanges = StateRanges()
+end
+
+function DiffCache(p_non_diff::ParametersNonDiff)
+    (; basin, pump, outlet, pid_control) = p_non_diff
+    n_basin = length(basin.node_id)
+    return (;
+        current_storage = zeros(n_basin),
+        current_low_storage_factor = zeros(n_basin),
+        current_level = zeros(n_basin),
+        current_area = zeros(n_basin),
+        current_cumulative_precipitation = zeros(n_basin),
+        current_cumulative_drainage = zeros(n_basin),
+        flow_rate_pump = zeros(length(pump.node_id)),
+        flow_rate_outlet = zeros(length(outlet.node_id)),
+        error_pid_control = zeros(length(pid_control.node_id)),
+    )
+end
+
+@kwdef struct Parameters{C1, C2, C3, C4, C5, T}
+    p_non_diff::ParametersNonDiff{C1, C2, C3, C4, C5}
+    diff_cache::DiffCache{T} = DiffCache(p_non_diff)
+    p_mutable::ParametersMutable = ParametersMutable()
+end
```

Comment:
Also for these structs a short docstring would be helpful to the reader. I know we should've added some before.